game-core/src/main/java/games/strategy/engine/framework/GameRunner.java

```diff
@@ -75,36 +76,54 @@
 import games.strategy.util.Interruptibles;
 import games.strategy.util.Version;
 import javafx.application.Application;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
 import lombok.extern.java.Log;
 
 /**
  * GameRunner - The entrance class with the main method.
  * In this class commonly used constants are getting defined and the Game is being launched
  */
 @Log
-public class GameRunner {
-
+public final class GameRunner {
   public static final String TRIPLEA_HEADLESS = "triplea.headless";
-
-  // not arguments:
   public static final int PORT = 3300;
-
   public static final int MINIMUM_CLIENT_GAMEDATA_LOAD_GRACE_TIME = 20;
 
   private static final GameSelectorModel gameSelectorModel = new GameSelectorModel();
   private static final SetupPanelModel setupPanelModel = new SetupPanelModel(gameSelectorModel);
+  private static final AtomicReference<Context> contextRef = new AtomicReference<>();
   private static JFrame mainFrame;
 
+  private GameRunner() {}
+
+  private static Context getContext() {
+    final @Nullable Context context = contextRef.get();
+    checkState(
+        context != null,
+        "GameRunner execution context has not been initialized. Did you forget to call GameRunner#start()?");
+    return context;
+  }
+
   /**
-   * Launches the "main" TripleA gui enabled game client.
-   * No args will launch a client, additional args can be supplied to specify additional behavior.
-   * Warning: game engine code invokes this method to spawn new game clients.
+   * Starts a new UI-enabled game client. This method will return before the game client UI exits. The game client UI
+   * will continue to run until it is shut down by the user.
+   *
+   * <p>
+   * Command-line arguments are specified via {@code context}. No arguments will launch a client; additional arguments
+   * can be supplied to specify additional behavior.
+   * </p>
+   *
+   * @throws IllegalStateException If called from a headless environment.
    */
-  public static void main(final String[] args) {
-    Preconditions.checkState(!GraphicsEnvironment.isHeadless(),
-        "UI client launcher invoked from headless environment. This is current prohibited by design to "
+  public static void start(final Context context) {
+    checkNotNull(context);
+    checkState(!GraphicsEnvironment.isHeadless(),
+        "UI client launcher invoked from headless environment. This is currently prohibited by design to "
             + "avoid UI rendering errors in the headless environment.");
 
+    contextRef.set(context);
```

Comment:
One last question: What benefit do you expect from using an atomic reference?
Or is it simply being used as a wrapper?

Reply:
The driving requirement here is that accessing the `Context` reference be thread-safe because I have no idea on which threads the lobby client may call `joinGame()` and `hostGame()`.  (I can guarantee they are not the same thread that called `GameRunner#start()` because that thread will have terminated by the time a lobby client is running.)  In addition, other uses of `Context` may introduce other callers on other threads who need to have access to it.

I originally used a `volatile Context` field instead of a `final AtomicReference<Context>` to achieve the same thing.  I thought the `AtomicReference` was cleaner, but that's debatable.